src/renderers/WebGLRenderer.js

```diff
@@ -268,27 +268,28 @@ function WebGLRenderer( parameters ) {
 
 		extensions.get( 'OES_texture_float_linear' );
 
-		utils = new WebGLUtils( _gl, extensions, capabilities );
+		utils = parameters.utils !== undefined ? parameters.utils : new WebGLUtils( _gl, extensions, capabilities );
```

Comment:
Why have you changed these lines in `initGLContext()`? That does not look right to me.

Reply:
This makes it possible to use custom WebGLXxx helpers with builtin WebGLRenderer. Currently the only way to customize renderer is to write a new one entirely from scratch.

Reply:
This PR is only about introducing `WebGLMaterials`, nothing else. Please don't put changes into the PR which are not approved in #19292. Otherwise this PR will never get merged.

So please focus on introducing `WebGLMaterials`. Extending the semantics of `parameters` like this will not be merged since this object is intended for creating the WebGL rendering context only. Besides, it do not vote to make the internal classes of `WebGLRenderer` exchangeable. This classes are for internal use only and no part of the public API. If we would expose them, user could rely on exchanging components although the project will change, replace or remove them over time.

I've mentioned this already multiple times but I do not vote to open-up the renderer in this way.

Reply:
ok, I will split this into 2 PRs

Reply:
Thanks! 👍

Reply:
Is this a pattern similar to https://github.com/mrdoob/three.js/pull/15490 ?

I feel that a lot of flexibility could be introduced with this pattern.

Reply:
@Mugen87 i think this is the same approach present in some cases and proposed in others. Would it make sense to document these somewhere. I believe that this is similar to `depthMaterial` or whatever it was, that wasn't a documented property but would be consumed by the renderer if present (even if not by design, this is how it worked). 

The buffer proposal is very similar, the difference i think is that what's proposed here is a very complex object above GL, while something like a render buffer is a direct GL primitive.

Reply:
Sorry, I don't understand. This change is totally unrelated to your PR. It just shifts code from one place to another.